File: src/main/java/com/github/nicoalvarezz/core/MasterWorker.java
package com.github.nicoalvarezz.core;

import com.github.nicoalvarezz.config.ConfigManager;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.nio.channels.ServerSocketChannel;
import java.nio.channels.SocketChannel;

public class MasterWorker {
    private final ServerSocketChannel serverChannel;
    private final ConfigManager config;
    private volatile boolean running;
//    private final Selector selector;

    public MasterWorker(ConfigManager config) throws IOException {
        this.config = config;
        this.serverChannel = ServerSocketChannel.open();
        this.serverChannel.bind(new InetSocketAddress(config.getPort()));
        this.serverChannel.configureBlocking(false);
//        this.selector = Selector.open();
//        this.serverChannel.register(selector, SelectionKey.OP_ACCEPT);
    }

    public void start() {
        Thread.startVirtualThread(() -> {
            while (running) {
                try {
                    SocketChannel clientChannel = serverChannel.accept();
                    Thread.startVirtualThread(() -> handelConnection(clientChannel));
                } catch (IOException e) {
                    if (running) {
                        // log error and continue
                    }
                }
            }
        });
    }

    private void handelConnection(SocketChannel clientChannel) {
        // we create a new virtual worker per new request
        VirtualWorker worker = new VirtualWorker(clientChannel);
        worker.run();
    }
}

File: app/src/main/kotlin/com/github/nicoalvarezz/core/MasterWorker.kt
package com.github.nicoalvarezz.core

import java.net.InetSocketAddress
import java.nio.channels.ServerSocketChannel
import java.nio.channels.SocketChannel
import java.util.concurrent.atomic.AtomicBoolean

class MasterWorker(config: ConfigurationManager) {
    private val serverChannel: ServerSocketChannel = ServerSocketChannel.open();
    private val isRunning = AtomicBoolean(true)

    init {
        val proxyConfig = config.getConfig();
        serverChannel.apply {
            bind(InetSocketAddress(proxyConfig.host, proxyConfig.port), 50)
            configureBlocking(true) // Configuring blocking IO since we are using virtual threads
        }
    }

    fun start() {
        println("Starting Master Worker on ${serverChannel.localAddress}")

        while (isRunning.get()) {
            try {
                val clientChannel = serverChannel.accept()
                handleNewConnection(clientChannel)
            } catch (e: Exception) {
                println("Error accepting new connection: ${e.message}")
            }
        }
    }

    fun shutdown() {
        isRunning.set(false);
        serverChannel.close()

        // gracefully shut down active virtual workers
    }

    fun getLocalAddress() = serverChannel.localAddress

    private fun handleNewConnection(clientChannel: SocketChannel) {
        // create a new virtual thread to handle new connection
        println("Connected: ${clientChannel.remoteAddress}")
    }
}